SharpSploit/Execution/Injection/Execution.cs

```diff
@@ -0,0 +1,267 @@
+﻿using System;
+using System.Linq;
+using System.Reflection;
+using System.Diagnostics;
+
+namespace SharpSploit.Execution.Injection
+{
+    /// <summary>
+    /// Base class for Injection strategies.
+    /// </summary>
+    public abstract class ExecutionTechnique
+    {
+
+        //An array containing a set of PayloadType objects that are supported.
+        protected Type[] supportedPayloads;
+
+        /// <summary>
+        /// Informs objects using this technique whether or not it supports the type of a particular payload.
+        /// </summary>
+        /// <author>The Wover (@TheRealWover)</author>
+        /// <param name="payload">A payload.</param>
+        /// <returns>Whether or not the payload is of a supported type for this strategy.</returns>
+        public abstract bool IsSupportedPayloadType(PayloadType payload);
+
+        /// <summary>
+        /// Internal method for setting the supported payload types. Used in constructors.
+        /// </summary>
+        /// <author>The Wover (@TheRealWover)</author>
+        abstract internal void DefineSupportedPayloadTypes();
+
+        /// <summary>
+        /// Inject and execute a payload in the target process using a specific allocation technique.
+        /// </summary>
+        /// <author>The Wover (@TheRealWover)</author>
+        /// <param name="Payload">The type of payload to execute.</param>
+        /// <param name="AllocationTechnique">The allocation technique to use.</param>
+        /// <param name="Process">The target process.</param>
+        /// <returns>bool</returns>
+        public bool Inject(PayloadType Payload, AllocationTechnique AllocationTechnique, Process Process)
```

Comment:
I've changed several of these sorts of functions to be virtual, since they are almost all being non-explicitly hidden in inherited classes (RemoteThreadCreate in this case). I'm a little confused why this Inject function, however, can't seem to be marked as virtual and overriden in the inherited class (RemoteThreadCreate), this changes the functionality to always throw an unsupported payloadtype exception.

Keeping this as-is works, at least in the case of the InjectionTests as currently written. I worry that this will cause RemoteThreadCreate to fail if someone has a reference to an object specifically typed as RemoteThreadCreate, instead of a generic type of ExecutionTechnique.

Reply:
Yeah, so this stuff is confusing. There needs to be an overload that takes PayloadType in order for the polymorphism to work. I tried out having it as an abstract function, but then that forced every subclass to provide the same logic: If the payload type is not supported, return an error, otherwise call the appropriate function. Because they all had to implement the function with the same logic, I saw it as unnecessary code duplication, so I implemented it in the parent class in this way.

I understand your concern about how runtime types call the appropriate function. That is why I used the `this` object. The end result is that: if an Inject method is called for which there is no overload specifying the specific Type of the payload, then the base function in the parent class will be called. At this point, it would probably be safe to assume that there is no function supporting what the user requested. But we check anyway and attempt to redirect to it if it exists. If not, then we throw the PayloadTypeNotSupported exception. This way, we cover every case that could occur, and throw an exception specific to this situation if an error occurs.